Fix the data dynamic property in WP_Term

anton-vlasenko opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

This PR aims to solve the issue with the `WP_Term::$data` dynamic property. 
The `WP_Term` class employs the `__get` magic method to compute the object data.
However, since PHP 8.2 does not support dynamic properties, it is better to eliminate this approach.

Trac ticket: https://core.trac.wordpress.org/ticket/58087

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

mukeshpanchal27 commented on src/wp-includes/class-wp-term.php:
```diff
@@ -226,20 +226,54 @@ public function to_array() {
+				'WP_Term::$data',
+				'6.4.0',
+				sprintf(
+				/* translators: %s: The method name to be used instead of the WP_Term::$data property. */
+					__( 'Use %s instead.' ),
```
```suggestion
					/* translators: %s: The method name to be used instead of the WP_Term::$data property. */
					__( 'Use %s instead.' ),
```

mukeshpanchal27 commented on tests/phpunit/tests/term.php:
```diff
@@ -308,4 +308,87 @@ public function test_orphan_category() {
+
+	/**
+	 * @ticket 58087
+	 * @coversNothing
+	 *
+	 * @param stdClass $actual An instance of stdClass containing the values to initialize the WP_Term object.
+	 * @param array $expected An array containing the expected values for the WP_Term object.
+	 *
+	 * @dataProvider data_get_raw_data_should_return_correct_values
```
```suggestion
	 * @dataProvider data_get_raw_data_should_return_correct_values
	 *
	 * @param stdClass $actual An instance of stdClass containing the values to initialize the WP_Term object.
	 * @param array $expected An array containing the expected values for the WP_Term object.
```

anton-vlasenko replied:
Thank you for the review, @mukeshpanchal27.
I agree with the proposed changes; however, I don't think the `@covers` tag should be removed.
Fixed in 0e8df1a1a6d64170a8b97e193f49d48571e23039

anton-vlasenko committed:
Solves the issue with the WP_Term::$data dynamic property.
The WP_Term class employs the __get magic method to compute the object data.
However, since PHP 8.2 does not support dynamic properties, it is better to eliminate this approach.

anton-vlasenko committed:
Update src/wp-includes/class-wp-term.php

Fix indentation.

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

anton-vlasenko committed:
Reorder tags per https://github.com/WordPress/wordpress-develop/pull/4307#discussion_r1332553806

anton-vlasenko committed:
Reorder docblock tags.

anton-vlasenko committed:
Move the test method for WP_Term::get_raw_data() into a separate Test Class.

anton-vlasenko commented:
Thank you for the code review, @mukeshpanchal27.
I have resolved the merge conflict and addressed the feedback from the code review.
The PR is now ready for consideration.

anton-vlasenko committed:
Update the WordPress version to 6.5.0.

anton-vlasenko commented:
I'm closing this PR as the solution needs [more refinement](https://core.trac.wordpress.org/ticket/58087#comment:28).

hellofromtonya commented:
Reopening this PR as IMO it's the most simplistic approach to resolve the issue and goals ([see my reasonings here](https://core.trac.wordpress.org/ticket/58087#comment:39)).

hellofromtonya committed:
Code quality improvements.

A few code quality improvements:

* Expanded docs.
* Renaming of variable to align to what they are: db columns.

hellofromtonya commented:
Reclosing this PR. Why? Discovered today that Core adds/uses a lot of dynamic properties for `WP_Term`. Thus, it no longer makes sense to only deal with the `data` property. A broader solution will be needed for all of the dynamic properties.
